credoai/modules/model_modules/security.py

```diff
@@ -0,0 +1,145 @@
+        self.model = model.model
+        self.victim_model = SklearnClassifier(self.model)
+        np.random.seed(10)
+
+    def run(self):
+        """Runs the assessment process
+
+        Returns
+        -------
+        dict
+            Key: metric name
+            Value: metric value
+        """
+        extraction_performance = self._extraction_attack()
+
+        attack_scores = {
+            **extraction_performance
+        }
+
+        self.results = attack_scores
+
+        return self
+
+    def prepare_results(self):
+        """Prepares results for export to Credo AI's Governance App
+
+        Structures a subset of results for export as a dataframe with appropriate structure
+        for exporting. See credoai.modules.credo_module.
+
+        Returns
+        -------
+        pd.Series
+
+        Raises
+        ------
+        NotRunError
+            If results have not been run, raise
+        """
+        if self.results is not None:
+            return pd.Series(self.results, name='value')
+        else:
+            raise NotRunError("Results not created yet. Call 'run' to create results")
+
+    def _extraction_attack(self):
+        """Model extraction security attack
+
+        In model extraction, the adversary only has access to the prediction API of a target model
+            which she queries to extract information about the model internals and train a substitute model.
+
+        Returns
+        -------
+        dict
+            Key: extraction_attack_score
+            Value: accuracy of the thieved model / accuracy of the victim model, corrected for chance
+        """
+        # use half of the test data for model extraction and half for evaluation
+        len_steal = int(len(self.x_test)/2)
+        indices = np.random.permutation(len(self.x_test))
+        x_steal = self.x_test[indices[:len_steal]]
+        y_steal = self.y_test[indices[:len_steal]]
+        x_test = self.x_test[indices[len_steal:]]
+        y_test = self.y_test[indices[len_steal:]]
+
+        # extract
+        copycat = CopycatCNN(
+            classifier=self.victim_model,
+            nb_epochs=5,
+            nb_stolen=len_steal
+            )
+
+        thieved_model = self._get_model(x_steal.shape[1])
+        thieved_classifier = KerasClassifier(thieved_model)
+
+        thieved_classifier = copycat.extract(x_steal, thieved_classifier=thieved_classifier)
+
+        # evaluate 
+        y_true = [np.argmax(y, axis=None, out=None) for y in y_test]
+
+        y_pred = [np.argmax(y, axis=None, out=None) for y in thieved_classifier._model.predict(x_test)]
+        thieved_classifier_acc = sk_metrics.accuracy_score(y_true, y_pred)
+
+        y_pred = self.victim_model._model.predict(x_test)
+        victim_classifier_acc = sk_metrics.accuracy_score(y_true, y_pred)
+
+        metrics = {
+            'extraction_attack_score': max((thieved_classifier_acc - 0.5) / (victim_classifier_acc - 0.5), 0)
```

Comment:
Nit: Are you organizing this as a dictionary because you intend (or there is the potential) to add other scores in the future?

Reply:
Correct. ART for example already has two other extraction methods that can be added. And of course there are other performance metrics we can add.